// validate-any/dist/classes/Validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Validator {
    static WRONG_TYPE = `Value is not of the correct type`;
    static WRONG_VALUE = `Value is not allowed`;
    schema = "";
    /**
     * Schema of the validator as a JSON object
     *
     * @returns JSON Schema of the object
     */
    getSchema() {
        return JSON.parse(this.schema);
    }
    success(data) {
        return {
            success: true,
            errors: [],
            data: data
        };
    }
    failure(locator, message, validator, data) {
        return {
            success: false,
            errors: [
                {
                    location: locator.getLocation(),
                    message,
                    expected: validator.getSchema(),
                    value: data
                }
            ],
            data: undefined
        };
    }
    throw(locator, message) {
        throw new Error(locator.getLocation() + ": " + message);
    }
}
exports.default = Validator;
