// ytmusic-api/dist/interfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PLAYLIST_FULL = exports.ALBUM_FULL = exports.ARTIST_FULL = exports.VIDEO_FULL = exports.SONG_FULL = exports.ALBUM_DETAILED = exports.ARTIST_DETAILED = exports.VIDEO_DETAILED = exports.SONG_DETAILED = exports.ALBUM_BASIC = exports.ARTIST_BASIC = exports.THUMBNAIL_FULL = void 0;
const validate_any_1 = require("validate-any");
exports.THUMBNAIL_FULL = (0, validate_any_1.OBJECT)({
    url: (0, validate_any_1.STRING)(),
    width: (0, validate_any_1.NUMBER)(),
    height: (0, validate_any_1.NUMBER)()
});
exports.ARTIST_BASIC = (0, validate_any_1.OBJECT)({
    artistId: (0, validate_any_1.STRING)(),
    name: (0, validate_any_1.STRING)()
});
exports.ALBUM_BASIC = (0, validate_any_1.OBJECT)({
    albumId: (0, validate_any_1.STRING)(),
    name: (0, validate_any_1.STRING)()
});
exports.SONG_DETAILED = (0, validate_any_1.OBJECT)({
    type: (0, validate_any_1.STRING)("SONG"),
    videoId: (0, validate_any_1.STRING)(),
    name: (0, validate_any_1.STRING)(),
    artists: (0, validate_any_1.LIST)(exports.ARTIST_BASIC),
    album: exports.ALBUM_BASIC,
    duration: (0, validate_any_1.NUMBER)(),
    thumbnails: (0, validate_any_1.LIST)(exports.THUMBNAIL_FULL)
});
exports.VIDEO_DETAILED = (0, validate_any_1.OBJECT)({
    type: (0, validate_any_1.STRING)("VIDEO"),
    videoId: (0, validate_any_1.STRING)(),
    name: (0, validate_any_1.STRING)(),
    artists: (0, validate_any_1.LIST)(exports.ARTIST_BASIC),
    duration: (0, validate_any_1.NUMBER)(),
    thumbnails: (0, validate_any_1.LIST)(exports.THUMBNAIL_FULL)
});
exports.ARTIST_DETAILED = (0, validate_any_1.OBJECT)({
    artistId: (0, validate_any_1.STRING)(),
    name: (0, validate_any_1.STRING)(),
    type: (0, validate_any_1.STRING)("ARTIST"),
    thumbnails: (0, validate_any_1.LIST)(exports.THUMBNAIL_FULL)
});
exports.ALBUM_DETAILED = (0, validate_any_1.OBJECT)({
    type: (0, validate_any_1.STRING)("ALBUM"),
    albumId: (0, validate_any_1.STRING)(),
    playlistId: (0, validate_any_1.STRING)(),
    name: (0, validate_any_1.STRING)(),
    artists: (0, validate_any_1.LIST)(exports.ARTIST_BASIC),
    year: (0, validate_any_1.OR)((0, validate_any_1.NUMBER)(), (0, validate_any_1.NULL)()),
    thumbnails: (0, validate_any_1.LIST)(exports.THUMBNAIL_FULL)
});
exports.SONG_FULL = (0, validate_any_1.OBJECT)({
    type: (0, validate_any_1.STRING)("SONG"),
    videoId: (0, validate_any_1.STRING)(),
    name: (0, validate_any_1.STRING)(),
    artists: (0, validate_any_1.LIST)(exports.ARTIST_BASIC),
    duration: (0, validate_any_1.NUMBER)(),
    thumbnails: (0, validate_any_1.LIST)(exports.THUMBNAIL_FULL),
    description: (0, validate_any_1.STRING)(),
    formats: (0, validate_any_1.LIST)((0, validate_any_1.OBJECT)()),
    adaptiveFormats: (0, validate_any_1.LIST)((0, validate_any_1.OBJECT)())
});
exports.VIDEO_FULL = (0, validate_any_1.OBJECT)({
    type: (0, validate_any_1.STRING)("VIDEO"),
    videoId: (0, validate_any_1.STRING)(),
    name: (0, validate_any_1.STRING)(),
    artists: (0, validate_any_1.LIST)(exports.ARTIST_BASIC),
    duration: (0, validate_any_1.NUMBER)(),
    thumbnails: (0, validate_any_1.LIST)(exports.THUMBNAIL_FULL),
    description: (0, validate_any_1.STRING)(),
    unlisted: (0, validate_any_1.BOOLEAN)(),
    familySafe: (0, validate_any_1.BOOLEAN)(),
    paid: (0, validate_any_1.BOOLEAN)(),
    tags: (0, validate_any_1.LIST)((0, validate_any_1.STRING)())
});
exports.ARTIST_FULL = (0, validate_any_1.OBJECT)({
    artistId: (0, validate_any_1.STRING)(),
    name: (0, validate_any_1.STRING)(),
    type: (0, validate_any_1.STRING)("ARTIST"),
    thumbnails: (0, validate_any_1.LIST)(exports.THUMBNAIL_FULL),
    description: (0, validate_any_1.STRING)(),
    topSongs: (0, validate_any_1.LIST)((0, validate_any_1.OBJECT)({
        type: (0, validate_any_1.STRING)("SONG"),
        videoId: (0, validate_any_1.STRING)(),
        name: (0, validate_any_1.STRING)(),
        artists: (0, validate_any_1.LIST)(exports.ARTIST_BASIC),
        album: exports.ALBUM_BASIC,
        thumbnails: (0, validate_any_1.LIST)(exports.THUMBNAIL_FULL)
    })),
    topAlbums: (0, validate_any_1.LIST)(exports.ALBUM_DETAILED)
});
exports.ALBUM_FULL = (0, validate_any_1.OBJECT)({
    type: (0, validate_any_1.STRING)("ALBUM"),
    albumId: (0, validate_any_1.STRING)(),
    playlistId: (0, validate_any_1.STRING)(),
    name: (0, validate_any_1.STRING)(),
    artists: (0, validate_any_1.LIST)(exports.ARTIST_BASIC),
    year: (0, validate_any_1.OR)((0, validate_any_1.NUMBER)(), (0, validate_any_1.NULL)()),
    thumbnails: (0, validate_any_1.LIST)(exports.THUMBNAIL_FULL),
    description: (0, validate_any_1.STRING)(),
    songs: (0, validate_any_1.LIST)(exports.SONG_DETAILED)
});
exports.PLAYLIST_FULL = (0, validate_any_1.OBJECT)({
    type: (0, validate_any_1.STRING)("PLAYLIST"),
    playlistId: (0, validate_any_1.STRING)(),
    name: (0, validate_any_1.STRING)(),
    artist: exports.ARTIST_BASIC,
    videoCount: (0, validate_any_1.NUMBER)(),
    thumbnails: (0, validate_any_1.LIST)(exports.THUMBNAIL_FULL)
});
